opm/simulators/flow/RFTContainer.cpp

```diff
@@ -0,0 +1,234 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#include <config.h>
+#include <opm/simulators/flow/RFTContainer.hpp>
+
+#include <opm/grid/common/CommunicationUtils.hpp>
+
+#include <opm/input/eclipse/EclipseState/EclipseState.hpp>
+#include <opm/input/eclipse/Schedule/RFTConfig.hpp>
+#include <opm/input/eclipse/Schedule/Schedule.hpp>
+#include <opm/input/eclipse/Schedule/Well/WellConnections.hpp>
+
+#include <opm/material/fluidsystems/BlackOilDefaultIndexTraits.hpp>
+#include <opm/material/fluidsystems/BlackOilFluidSystem.hpp>
+#include <opm/material/fluidsystems/GenericOilGasWaterFluidSystem.hpp>
+
+#include <opm/output/data/Wells.hpp>
+
+#include <algorithm>
+#include <tuple>
+
+namespace {
+
+template<class Scalar>
+void gatherAndUpdateMap(std::map<std::size_t, Scalar>& local_map,
+                        const Opm::Parallel::Communication& comm)
+{
+    std::vector<std::pair<int, Scalar>> pairs(local_map.begin(), local_map.end());
+    std::vector<std::pair<int, Scalar>> all_pairs;
+    std::vector<int> offsets;
+
+    std::tie(all_pairs, offsets) = Opm::allGatherv(pairs, comm);
+
+    // Update maps on all ranks
+    for (auto i = static_cast<std::size_t>(offsets[0]); i < all_pairs.size(); ++i) {
+        const auto& key_value = all_pairs[i];
+        if (auto candidate = local_map.find(key_value.first); candidate != local_map.end()) {
+            const Scalar prev_value = candidate->second;
+            candidate->second = std::max(prev_value, key_value.second);
+        }
+        else {
+            local_map[key_value.first] = key_value.second;
+        }
+    }
+}
+
+}
+
+namespace Opm {
+
+template<class FluidSystem>
+void RFTContainer<FluidSystem>::
+addToWells(data::Wells& wellDatas,
+           const std::size_t reportStepNum,
+           const Parallel::Communication& comm)
+{
+    if (comm.size() > 1) {
+        gatherAndUpdateMap(oilConnectionPressures_, comm);
+        gatherAndUpdateMap(waterConnectionSaturations_, comm);
+        gatherAndUpdateMap(gasConnectionSaturations_, comm);
+    }
+
+    const auto& rft_config = schedule_[reportStepNum].rft_config();
+    for (const auto& well: schedule_.getWells(reportStepNum)) {
+
+        // don't bother with wells not on this process
+        if (!wellQuery_(well.name())) {
+            continue;
+        }
+
+        //add data infrastructure for shut wells
+        if (!wellDatas.count(well.name())) {
+            data::Well wellData;
+
+            if (!rft_config.active()) {
+                continue;
+            }
```

Comment:
I suppose we could move that check ahead of the loop, and **possibly** ahead of the `comm.size()` check as well.

Reply:
indeed